File: core-py/examples/resilient_client.py
# /// script
# requires-python = ">=3.11"
# dependencies = ["oigtl"]
# ///
"""Demonstration of ``oigtl.net.Client`` resilience features.

Ports ``core-cpp/examples/resilient_client.cpp`` to Python. Runs
server + client in one process so the demo is self-contained. We
deliberately tear the server down mid-session to force
auto-reconnect, then restart on the same port so buffered
messages drain.

Run::

    uv run --project core-py python core-py/examples/resilient_client.py

Expected output (details may vary by timing):

    [client] connecting to 127.0.0.1:<port> ...
    [client] Phase 1: 3 messages while connected
      [server] TRANSFORM from demo-tracker: [0.00, 0.00, 0.00]
      [server] TRANSFORM from demo-tracker: [1.00, 2.00, 3.00]
      [server] TRANSFORM from demo-tracker: [2.00, 4.00, 6.00]
    [client] Phase 2: stopping server to induce drop
    [client] on_disconnected
    [client] Phase 2: 5 messages during outage
    [client] reconnect attempt 1 failed; next in ~200 ms
    [client] Phase 3: restart server
      [server] TRANSFORM from demo-tracker: [3.00, 6.00, 9.00]
      ...
    [client] on_connected (#2)
    [client] Phase 3: 2 post-reconnect messages
      [server] TRANSFORM from demo-tracker: [8.00, 16.00, 24.00]
      [server] TRANSFORM from demo-tracker: [9.00, 18.00, 27.00]

    SUMMARY:
      server session 1 received: 3 messages
      server session 2 received: 7 messages (5 buffered + 2 live)
      disconnects observed:      1
      reconnects observed:       1
    PASS
"""

from __future__ import annotations

import asyncio
import sys
from datetime import timedelta

from oigtl.messages import Transform
from oigtl.net import Client, ClientOptions, OfflineOverflow, Server


async def _run_server_for(port: int, counter: list[int]) -> Server:
    """Boot a server on *port*, bump *counter* for every TRANSFORM."""
    server = await Server.listen(port)

    @server.on(Transform)
    async def _(env, peer):
        counter[0] += 1
        tx = env.body.matrix[9:12]
        print(
            f"  [server] TRANSFORM from {env.header.device_name}: "
            f"[{tx[0]:.2f}, {tx[1]:.2f}, {tx[2]:.2f}]"
        )

    return server


def _tf(i: int) -> Transform:
    return Transform(matrix=[1, 0, 0, 0, 1, 0, 0, 0, 1,
                             float(i), float(i * 2), float(i * 3)])


async def main() -> int:
    rx1 = [0]
    rx2 = [0]

    # ---- Phase 1: boot initial server -------------------------------
    s1 = await _run_server_for(0, rx1)
    port = s1.port
    s1_task = asyncio.create_task(s1.serve())

    # ---- Configure resilient client ---------------------------------
    opt = ClientOptions(
        auto_reconnect=True,
        tcp_keepalive=True,
        offline_buffer_capacity=20,
        offline_overflow_policy=OfflineOverflow.DROP_OLDEST,
        reconnect_initial_backoff=timedelta(milliseconds=100),
        reconnect_max_backoff=timedelta(milliseconds=500),
        default_device="demo-tracker",
    )

    print(f"[client] connecting to 127.0.0.1:{port} ...")
    client = await Client.connect("127.0.0.1", port, opt)

    reconnects = 0
    drops = 0

    @client.on_connected
    def _():
        nonlocal reconnects
        reconnects += 1
        print(f"[client] on_connected (#{reconnects})")

    @client.on_disconnected
    def _(cause):
        nonlocal drops
        drops += 1
        print("[client] on_disconnected")

    @client.on_reconnect_failed
    def _(attempt, delay):
        ms = int(delay.total_seconds() * 1000)
        print(
            f"[client] reconnect attempt {attempt} failed; "
            f"next in ~{ms} ms"
        )

    # ---- Phase 1: happy path ---------------------------------------
    print("[client] Phase 1: 3 messages while connected")
    for i in range(3):
        await client.send(_tf(i))
        await asyncio.sleep(0.05)
    await asyncio.sleep(0.15)

    # ---- Phase 2: induce drop --------------------------------------
    print("[client] Phase 2: stopping server to induce drop")
    await s1.close()
    await s1_task

    print("[client] Phase 2: 5 messages during outage")
    for i in range(3, 8):
        try:
            await client.send(_tf(i))
        except Exception as e:
            print(f"[client] send {i} threw: {e}")
        await asyncio.sleep(0.05)

    # ---- Phase 3: restart server on same port ----------------------
    print(f"[client] Phase 3: restart server on port {port}")
    s2 = await _run_server_for(port, rx2)
    s2_task = asyncio.create_task(s2.serve())
    await asyncio.sleep(1.0)    # reconnect window

    print("[client] Phase 3: 2 post-reconnect messages")
    for i in range(8, 10):
        await client.send(_tf(i))
        await asyncio.sleep(0.1)
    await asyncio.sleep(0.3)

    # ---- Cleanup ---------------------------------------------------
    await client.close()
    await s2.close()
    await s2_task

    print(f"\nSUMMARY:")
    print(f"  server session 1 received: {rx1[0]} messages")
    print(f"  server session 2 received: {rx2[0]} messages")
    print(f"  disconnects observed:      {drops}")
    print(f"  reconnects observed:       {reconnects}")

    ok = (rx1[0] >= 3 and rx2[0] >= 2 and drops >= 1 and reconnects >= 1)
    print(f"\n{'PASS' if ok else 'FAIL (unexpected counts)'}")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: core-cpp/examples/resilient_client.cpp
// resilient_client.cpp — demonstrates oigtl::Client's resilience
// features end-to-end.
//
// Runs client + server in one process so the demo is
// self-contained. The server is deliberately dropped mid-session
// to force the client's auto_reconnect path, then restarted on
// the same port so the buffered messages drain.
//
// Build:
//   cmake --build core-cpp/build --target resilient_client
//   ./core-cpp/build/resilient_client
//
// Concept doc: core-cpp/CLIENT_GUIDE.md.

#include <atomic>
#include <chrono>
#include <cstdio>
#include <memory>
#include <mutex>
#include <thread>

#include "oigtl/client.hpp"
#include "oigtl/messages/status.hpp"
#include "oigtl/messages/transform.hpp"
#include "oigtl/server.hpp"

namespace msg = oigtl::messages;

// ----------------------------------------------------------------------
// ServerSession — owns an accepted peer and a reader thread. The
// reader exits promptly when the peer is closed from the outside
// (via stop()) or when the client disconnects naturally. Lets
// main() drop the server-side view of the connection without
// waiting for TCP to notice a vanished listener.
// ----------------------------------------------------------------------
class ServerSession {
 public:
    ServerSession(std::uint16_t port, std::atomic<int>& counter)
        : server_(oigtl::Server::listen(port, make_opt())),
          port_(server_.local_port()),
          counter_(counter) {
        reader_ = std::thread([this] { run(); });
    }

    ~ServerSession() {
        stop();
        if (reader_.joinable()) reader_.join();
    }

    std::uint16_t port() const { return port_; }

    void stop() {
        stop_requested_.store(true);
        // Close the server's accept socket (prevents future
        // accepts and breaks the current one, if any).
        server_.close();
        // Close the current peer so its receive_any unblocks.
        std::lock_guard<std::mutex> lk(peer_mu_);
        if (peer_) { try { peer_->close(); } catch (...) {} }
    }

 private:
    static oigtl::ServerOptions make_opt() {
        oigtl::ServerOptions o;
        // Short receive_timeout on the accepted peer so its
        // receive_any loop ticks and can notice stop().
        o.client_defaults.receive_timeout =
            std::chrono::milliseconds(200);
        return o;
    }

    void run() {
        try {
            auto p = server_.accept();
            {
                std::lock_guard<std::mutex> lk(peer_mu_);
                peer_ = std::make_unique<oigtl::Client>(std::move(p));
            }
            while (!stop_requested_.load()) {
                try {
                    auto inc = peer_->receive_any();
                    ++counter_;
                    if (inc.header.type_id == "TRANSFORM") {
                        auto env = oigtl::unpack_envelope<msg::Transform>(inc);
                        std::printf(
                            "  [server] TRANSFORM from %s: "
                            "[%.2f, %.2f, %.2f]\n",
                            inc.header.device_name.c_str(),
                            env.body.matrix[9],    // tx
                            env.body.matrix[10],   // ty
                            env.body.matrix[11]);  // tz
                    }
                } catch (const oigtl::transport::TimeoutError&) {
                    continue;    // tick and re-check stop_requested
                }
            }
        } catch (const std::exception& e) {
            std::printf("  [server] session ended: %s\n", e.what());
        }
    }

    oigtl::Server server_;
    std::uint16_t port_;
    std::atomic<int>& counter_;
    std::thread reader_;
    std::atomic<bool> stop_requested_{false};
    std::mutex peer_mu_;
    std::unique_ptr<oigtl::Client> peer_;
};

int main() {
    std::atomic<int> rx1{0};
    std::atomic<int> rx2{0};

    // Phase 1 server on a random free port. `std::uint16_t{0}`
    // rather than bare 0 so MSVC's /W4 doesn't warn about
    // narrowing the int literal through make_unique's perfect
    // forwarding.
    auto s1 = std::make_unique<ServerSession>(std::uint16_t{0}, rx1);
    const auto port = s1->port();

    oigtl::ClientOptions opt;
    opt.auto_reconnect          = true;
    opt.tcp_keepalive           = true;
    opt.offline_buffer_capacity = 20;
    opt.offline_overflow_policy =
        oigtl::ClientOptions::OfflineOverflow::DropOldest;
    opt.reconnect_initial_backoff = std::chrono::milliseconds(100);
    opt.reconnect_max_backoff     = std::chrono::milliseconds(500);
    opt.receive_timeout           = std::chrono::milliseconds(200);
    opt.default_device            = "demo-tracker";

    std::printf("[client] connecting to 127.0.0.1:%u ...\n", port);
    auto client = oigtl::Client::connect("127.0.0.1", port, opt);

    std::atomic<int> reconnects{0};
    std::atomic<int> drops{0};
    client.on_connected([&] {
        std::printf("[client] on_connected (#%d)\n", ++reconnects);
    });
    client.on_disconnected([&](std::exception_ptr) {
        std::printf("[client] on_disconnected\n");
        ++drops;
    });
    client.on_reconnect_failed(
        [&](int attempt, std::chrono::milliseconds delay) {
            std::printf(
                "[client] reconnect attempt %d failed; "
                "next in %lld ms\n",
                attempt,
                static_cast<long long>(delay.count()));
        });

    auto send_tf = [&](float tx, float ty, float tz) {
        msg::Transform m;
        // Column-major 3x4: R11..R33 then TX, TY, TZ.
        m.matrix = { 1,0,0, 0,1,0, 0,0,1, tx, ty, tz };
        client.send(m);
    };

    // ---- Phase 1: happy path --------------------------------------
    std::printf("[client] Phase 1: 3 messages while connected\n");
    for (int i = 0; i < 3; ++i) {
        send_tf(float(i), float(i * 2), float(i * 3));
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(150));

    // ---- Phase 2: induce drop -------------------------------------
    std::printf("[client] Phase 2: stopping server to induce drop\n");
    s1.reset();    // joins the reader thread; drops the peer

    std::printf("[client] Phase 2: 5 messages during outage\n");
    for (int i = 3; i < 8; ++i) {
        try {
            send_tf(float(i), float(i * 2), float(i * 3));
        } catch (const std::exception& e) {
            std::printf("[client] send %d threw: %s\n", i, e.what());
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

    // ---- Phase 3: restart server ----------------------------------
    std::printf("[client] Phase 3: restart server on port %u\n", port);
    auto s2 = std::make_unique<ServerSession>(port, rx2);
    std::this_thread::sleep_for(std::chrono::seconds(1));   // reconnect

    std::printf("[client] Phase 3: 2 post-reconnect messages\n");
    for (int i = 8; i < 10; ++i) {
        send_tf(float(i), float(i * 2), float(i * 3));
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(300));

    // ---- Cleanup --------------------------------------------------
    client.close();
    s2.reset();

    std::printf("\nSUMMARY:\n");
    std::printf("  server session 1 received: %d messages\n",
                rx1.load());
    std::printf("  server session 2 received: %d messages\n",
                rx2.load());
    std::printf("  disconnects observed:      %d\n", drops.load());
    std::printf("  reconnects observed:       %d\n", reconnects.load());

    const bool ok =
        rx1.load()        >= 3 &&
        rx2.load()        >= 2 &&
        drops.load()      >= 1 &&
        reconnects.load() >= 1;
    std::printf("\n%s\n", ok ? "PASS" : "FAIL (unexpected counts)");
    return ok ? 0 : 1;
}
